src/pal/src/misc/utils.cpp

```diff
@@ -323,3 +329,72 @@ void UTIL_SetLastErrorFromMach(kern_return_t MachReturn)
 }
 #endif //HAVE_VM_ALLOCATE
 
+#ifdef __APPLE__
+
+#define DARWIN_VERSION_MOJAVE 18
+
+/*++
+Function:
+  GetDarwinVersion() - Helper function to get the version of the Apple's darwin
+--*/
+static UINT GetDarwinVersion()
+{
+    static UINT version = 0;
+
+    if (!version)
+    {
+        char str[32] = {0};
+        size_t size = sizeof(str);
+        int mib[2];
+
+        mib[0] = CTL_KERN;
+        mib[1] = KERN_OSRELEASE;
+        int err = sysctl(mib, 2, str, &size, NULL, 0);
+        _ASSERTE(err == 0);
+        int cnt = sscanf(str, "%d", &version);
+        _ASSERTE(cnt == 1);
+        _ASSERTE(version > 0);
+    }
+
+    return version;
+}
+
+/*++
+Function:
+  IsRunningOnMojaveHardenedRuntime() - Test if the current process is running on Mojave hardened runtime
+--*/
+BOOL IsRunningOnMojaveHardenedRuntime()
+{
+    static int isRunningOnMojaveHardenedRuntime = -1;
+
+    if (isRunningOnMojaveHardenedRuntime == -1)
```

Comment:
Will this always be called from the same thread ?

Reply:
It doesn't matter. The worst case is that we would evaluate the stuff multiple times.

Reply:
It does matter. Note that the C/C++ spec allows compilers to duplicate the memory load and introduce a subtle race condition in the process. E.g. this can be transformed by the C/C++ compiler to:

```
static int isRunningOnMojaveHardenedRuntime = -1;

int result = isRunningOnMojaveHardenedRuntime;

if (isRunningOnMojaveHardenedRuntime == -1)
{
   ...
}

return result;
```